## Question 1

What's the version of NumPy that you installed?

In [48]:
import numpy as np
np.__version__

'1.21.5'

### Getting the data

In [49]:
!wget https://raw.githubusercontent.com/alexeygrigorev/mlbookcamp-code/master/chapter-02-car-price/data.csv

--2022-09-09 11:04:49--  https://raw.githubusercontent.com/alexeygrigorev/mlbookcamp-code/master/chapter-02-car-price/data.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.109.133, 185.199.108.133, 185.199.111.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.109.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1475504 (1.4M) [text/plain]
Saving to: ‘data.csv.1’

data.csv.1          100%[===================>]   1.41M  1.88MB/s    in 0.7s    

2022-09-09 11:04:50 (1.88 MB/s) - ‘data.csv.1’ saved [1475504/1475504]



## Question 2
How many records are in the dataset

In [50]:
import pandas as pd

data = pd.read_csv("data.csv")

In [51]:
data.shape

(11914, 16)

We have 11914 rows and 16 columns

## Question 3
Who are the most frequent car manufacturers (top-3) according to the dataset?

In [52]:
data.head()

Make       Model  Year             Engine Fuel Type  Engine HP  \
0  BMW  1 Series M  2011  premium unleaded (required)      335.0   
1  BMW    1 Series  2011  premium unleaded (required)      300.0   
2  BMW    1 Series  2011  premium unleaded (required)      300.0   
3  BMW    1 Series  2011  premium unleaded (required)      230.0   
4  BMW    1 Series  2011  premium unleaded (required)      230.0   

   Engine Cylinders Transmission Type     Driven_Wheels  Number of Doors  \
0               6.0            MANUAL  rear wheel drive              2.0   
1               6.0            MANUAL  rear wheel drive              2.0   
2               6.0            MANUAL  rear wheel drive              2.0   
3               6.0            MANUAL  rear wheel drive              2.0   
4               6.0            MANUAL  rear wheel drive              2.0   

                         Market Category Vehicle Size Vehicle Style  \
0  Factory Tuner,Luxury,High-Performance      Compact         Coupe   
1                     Luxury,Performance      Compact   Convertible   
2                Luxury,High-Performance      Compact         Coupe   
3                     Luxury,Performance      Compact         Coupe   
4                                 Luxury      Compact   Convertible   

   highway MPG  city mpg  Popularity   MSRP  
0           26        19        3916  46135  
1           28        19        3916  40650  
2           28        20        3916  36350  
3           28        18        3916  29450  
4           28        18        3916  34500

In [53]:
data['Make'].value_counts().head(n=3)

Chevrolet     1123
Ford           881
Volkswagen     809
Name: Make, dtype: int64

The most frequent car manufacturers are **Chevrolet, Ford, and Volkswagen**

## Qestion 4
What's the number of unique Audi car models in the dataset?

In [54]:
data[data['Make']=='Audi']['Model'].nunique()

34

We have **34 unique** models of Audi

## Question 5
How many columns in the dataset have missing values?

In [55]:
data.isnull().any(axis=0).sum()

5

We have 5 columns with missing values

## Question 6

In [56]:
print("1. Median value of \"Engine Cylinders\" ",data['Engine Cylinders'].median())

1. Median value of "Engine Cylinders"  6.0


In [57]:
print("2. Most frequent value of \"Engine Cylinders\" ", data['Engine Cylinders'].mode()[0])

2. Most frequent value of "Engine Cylinders"  4.0


In [58]:
data['Engine Cylinders'].fillna(data['Engine Cylinders'].mode()[0], inplace=True)

In [59]:
print("1. Median value of \"Engine Cylinders\" after mode fill",data['Engine Cylinders'].median())

1. Median value of "Engine Cylinders" after mode fill 6.0


The median value before and after the null values is the same.

## Question 7


1. Select all "Lotus" cars from the dataset.

In [144]:
lotus_data = data[data['Make'] == 'Lotus']

2. Select only columns "Engine HP", "Engine Cylinders".

In [145]:
lotus_data = lotus_data[["Engine HP", "Engine Cylinders"]]

In [146]:
lotus_data.shape

(29, 2)

3. Now drop all duplicated rows using drop_duplicates method (you should get a dataframe with 9 rows).

In [147]:
lotus_data.drop_duplicates(inplace=True)

In [148]:
lotus_data.shape

(9, 2)

4. Get the underlying NumPy array. Let's call it X

In [149]:
X = lotus_data.to_numpy()

5. Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX.

In [150]:
X_T = X.T

In [156]:
XTX = np.dot(X_T, X)

6. Invert **XTX**

In [157]:
XTX_inv = np.linalg.inv(XTX)

In [158]:
y = [1100, 800, 750, 850, 1300, 1000, 1000, 1300, 800]

In [159]:
XTX_inv

array([[ 5.53084235e-05, -2.94319825e-03],
       [-2.94319825e-03,  1.60588447e-01]])

In [160]:
w = np.dot(np.dot(XTX_inv, X.T), y)

In [161]:
w

array([  4.59494481, -63.56432501])

The value of the first element of w is **4.5949**